src/IO/ReadBufferFromEncryptedFile.cpp

```diff
@@ -67,8 +67,22 @@ off_t ReadBufferFromEncryptedFile::getPosition()
     return offset - available();
 }
 
+size_t ReadBufferFromEncryptedFile::getFileOffsetOfBufferEnd() const
+{
+    return offset;
+}
+
 bool ReadBufferFromEncryptedFile::nextImpl()
 {
+    Memory<> external_buffer;
+    std::optional<Buffer> old_buffer;
+    if (in->internalBuffer().empty())
```

Comment:
It is not normal to have empty internal buffer. In which case do you get it? With multiple layers of buffers it should have been passed from the first buffer in this layers chain under check `use_external_buffer`,
Please check how `use_external_buffer` argument is used in other buffers, it is better to do the same here. Also below in constructor it should be like in other buffers `: ReadBufferFromFileBase(buffer_size_, nullptr, 0)` -> `: ReadBufferFromFileBase( user_external_buffer ? 0 : buffer_size_, nullptr, 0)`. 
https://github.com/ClickHouse/ClickHouse/blob/484bf00c06ff9a6caefb8d67eab9d43332f4c6b1/src/IO/ReadBufferFromS3.cpp#L101
https://github.com/ClickHouse/ClickHouse/blob/484bf00c06ff9a6caefb8d67eab9d43332f4c6b1/src/Disks/IO/CachedOnDiskReadBufferFromFile.cpp#L55
https://github.com/ClickHouse/ClickHouse/blob/484bf00c06ff9a6caefb8d67eab9d43332f4c6b1/src/Disks/IO/ReadBufferFromRemoteFSGather.cpp#L51

Reply:
> I see that there are a lot of swap, but I wasn't sure that was the right approach to use here, because offsets from the encryption buffers don't map 1:1 to offsets in other buffers. 

ah, indeed, I see now.
So because `ReadBufferFromEncryptedFile` anyway copies the data in line 101, then we can just force `use_external_buffer = false` to `in`. May be just pass it as additional arg to lambda here https://github.com/ClickHouse/ClickHouse/blob/484bf00c06ff9a6caefb8d67eab9d43332f4c6b1/src/Disks/ObjectStorages/S3/S3ObjectStorage.cpp#L168 I like this way more because we will guarantee that buffer is allocated once and it seems just cleaner. wdyt?
The one thing which seems needed is to not allocate buffer for `ReadBufferFromEncryptedFile` in ctor for your case, because it will be unused afaik, because we will have 
```
ReadBufferFromRemoteFSGather -> CachedOnDIskReadBufferFromFile -> ReadBufferFromEncryptedFile -> ReadBufferFromS3
                            ^                                   ^                               ^
                            swap                               swap                             copy
```
e.g. because of this memory swap between `CachedOnDIskReadBufferFromFile` and `ReadBufferFromEncryptedFile`, memory allocated by `ReadBufferFromEncryptedFile` is unused.